repo-docs/release-process.md

```diff
@@ -0,0 +1,113 @@
+2. NPM Workspaces linking.
+   - NPM Workspaces creates a [symbolic link](https://en.wikipedia.org/wiki/Symbolic_link) between a package and its dependencies when they are part of the same repository. NPM Workspaces won't link packages if the `version` field of the dependency doesn't match the version specified in the `dependencies` field of the dependant package.
+3. Waterfall bumping.
+   - Bumping dependencies directly in `package.json` means that packages can determine whether they need to be bumped by just looking at their own `package.json`.
+
+On pre-releases, the first ten digits of the commit hash will be appended to the version like so:
+
+- `1.2.3` -> `1.2.3-pre.abcdef1234`
+
+Additionally, this task will update the `CHANGELOG.md` file of the package to contain the changes that were taken into account when bumping its version.
+
+This task does not make any changes to the `package-lock.json` file at the root of the repository, since doing so would cause [an error with NPM Workspaces](https://github.com/npm/cli/issues/5506).
+
+## `release:build` (build bumped packages)
+
+This task is run individually on every package, immediately after `release:phase1`.
+
+The purpose of this task is to re-build a project right after it was bumped and before it gets published. This is needed because many packages contain information about their own version in their compiled code.
+
+This task (and its dependencies) are automatically generated by `@coveo/nx-plugin`, which clones all `build` tasks and their respective dependencies. This is needed to ensure `release:build` depends on the current package's `release:phase1`. **Warning**: changes to `@coveo/nx-plugin` won't take effect until `npm install` is executed, which triggers Nx to rebuild its graph.
+
+## `release:phase2` (publish to NPM)
+
+This task is run individually on every package, immediately after `release:build`.
+
+This workflow works on the assumption that we can't bump a package's version if any of its dependencies isn't published to NPM. The purpose of this task is to publish packages to NPM before dependant packages bump their versions.
+
+If a package is already published to NPM, this task will exit without error. After a package is published, this task will repeatedly query NPM until it confirms that the package exists in the registry.
+
+## `release:phase3` (reify `package-lock.json`)
+
+This task is run once at the root of the repository after every package's version was bumped.
+
+The purpose of this task is to re-generate the root `package-lock.json` file using [Arborist](https://www.npmjs.com/package/@npmcli/arborist), which should confirm that the version bumps were successful.
+
+Manual releases will commit the new `package-lock.json` to the main branch, which ensures that contributors can run `npm install` without unnecessary changes being made to their `package-lock.json`.
```

Comment:
Isn't `npm ci` the one which installs based on `package-lock.json` without modifying it and `npm install` would read `package.json` and update deps which sometimes would update `package-lock.json`?

Reply:
Unfortunately, `npm install` also suffered from a broken `package-lock.json` when we used Lerna. We were forced to delete the whole `node_modules` before doing `npm install`, and then we'd get diffs between Lerna's `package-lock.json` and the local one.